pkgs/native_assets_cli/lib/src/api/link_config.dart

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:args/args.dart';
+import 'package:pub_semver/pub_semver.dart';
+
+import '../model/hook.dart';
+import '../model/hook_config.dart';
+import '../model/resource_identifiers.dart';
+import '../utils/map.dart';
+import 'architecture.dart';
+import 'asset.dart';
+import 'build_config.dart';
+import 'build_mode.dart';
+import 'hook_config.dart';
+import 'ios_sdk.dart';
+import 'os.dart';
+import 'resource.dart';
+
+part '../model/link_config.dart';
+
+/// The configuration for a link hook (`hook/link.dart`) invocation.
+///
+/// It consists of a subset of the fields from the [BuildConfig] already passed
+/// to the build hook, the [assets] from the build step, and the [resources]
+/// generated during the kernel compilation.
+abstract class LinkConfig implements HookConfig {
+  /// The list of assets to be linked. These are the assets generated by a
+  /// `build.dart` script destined for this packages `link.dart`.
+  List<Asset> get assets;
+
+  /// A collection of methods annotated with `@ResourceIdentifier`, which are
+  /// called in the tree-shaken Dart code. This information can be used to
+  /// dispose unused [assets].
+  ///
+  /// This is `null` in JIT mode, where no resources are collected.
+  List<Resource>? get resources;
```

Comment:
Ah interesting. We indeed don't want to give an empty list in JIT, because that would be ambigious with not having any resources reachable in AOT! Nice find!

Maybe it would be better if we actually surface a `DartBuildMode` with `DartBuildMode.jit` and `DartBuildMode.aot`, and which users can branch on. And then throw a `StateError` if users try to access this info.

Reply:
I like having a `DartBuildMode`, but if the only information is the availability of `resources`, then this feels like a complication. As for the `StateError`, I think a nullable field is more visible than a runtime error.

Reply:
For `null` users have to read the documentation of the getter. You won't be able to see on the call site that the null-check is to deal with the fact treeshaking is not enabled.

I'm thinking the call sites will be more readable if it is: `if(config.treeshakingEnabled) config.resources... else ...` or `if(config.dartBuildMode == DartBuildMode.aot) config.resources... else ...` instead of `config.resources?... ?? ...`.

Reply:
In that case I would rather rename the field, so that `if(config.treeshakingEnabled) ...` is just `treeshakingInformation?...`. Throwing a `StateError` is just setting a trap.

Reply:
It's a trap for the first time the user invokes the API (N=package author) but readability is more important (M=package user). M > N.

Yeah `if(config.treeshakingEnabled) treeshakingInformation... else ...` would probably be as good as `if(config.treeshakingInformation != null) treeshakingInformation... else ...`, so +1 for renaming.